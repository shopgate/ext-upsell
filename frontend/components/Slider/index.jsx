import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import getProductRelationsAction from '@shopgate/pwa-common-commerce/product/actions/getProductRelations';
import { getRelatedProductsFiltered } from '../../selectors';
import DefaultSlider from './components/DefaultSlider';
import getStyles from '../../styles/slider';

const styles = getStyles();

class Slider extends Component {
  static propTypes = {
    productId: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    headline: PropTypes.string,
    showName: PropTypes.bool,
    showPrice: PropTypes.bool,
  };

  static defaultProps = {
    headline: null,
    showName: false,
    showPrice: false,
  };

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(getProductRelationsAction({
      productId: this.props.productId,
      type: this.props.type,
    }));
  }

  render() {
    if (!this.props.products.length) {
      return null;
    }

    return (
      <div className={styles.wrapper}>
        {this.props.headline && <h3 className={styles.headline}>{this.props.headline}</h3> }
        <DefaultSlider
          products={this.props.products}
          showPrice={this.props.showPrice}
          showName={this.props.showName}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  products: getRelatedProductsFiltered({
    productId: props.productId,
    type: props.type,
  })(state),
});

export default connect(mapStateToProps)(Slider);
