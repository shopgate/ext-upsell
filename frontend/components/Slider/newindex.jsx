import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { useTheme } from '@shopgate/engage/core';
import { getRelatedProductsIdsByIdFiltered } from '../../selectors';
import getStyles from '../../styles/slider';

const styles = getStyles();

/**
 * Custom slider component
 * @param {string[]} relatedProductIds The ids of related products that are known to be in redux
 * @param {string} headline Headline for product slider
 * @return {JSK}
 */
const Slider = ({ relatedProductIds, headline, showPrice, showName }) => {
  const { ProductSlider } = useTheme();
  if (!relatedProductIds.length) {
    return null;
  }
  return (
    <div className={styles.wrapper}>
      {headline && <h3 className={styles.headline}>{headline}</h3> }
      <ProductSlider
        productIds={relatedProductIds}
        autoplay
        delay={7000}
        slidesPerView={2.3}
        snap={false}
        showPrice={showPrice}
        showName={showName}
      />
    </div>
  );
};

Slider.propTypes = {
  headLine: PropTypes.string,
  relatedProductIds: PropTypes.arrayOf(PropTypes.string),
};

Slider.defaultProps = {
  headLine: null,
  relatedProductIds: [],
};

/**
 * Returns products from redux.
 * @param {Object} state State.
 * @param {Object} props Props.
 * @returns {Object}
 */
const mapStateToProps = (state, props) => {
  const params = {
    productId: props.productId,
    type: props.type,
  };

  return {
    relatedProductIds: getRelatedProductsIdsByIdFiltered(params)(state),
  };
};

export default connect(mapStateToProps)(Slider);
