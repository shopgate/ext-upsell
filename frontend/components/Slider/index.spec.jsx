import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import Slider from './index';

const mockedStore = configureStore();

// eslint-disable-next-line react/prop-types, require-jsdoc
const MockedDummyComponent = props => (<div>{props.children}</div>);

jest.mock('./components/DefaultSlider', () => props => (
  <MockedDummyComponent {...props}>
    Mocked Slider
  </MockedDummyComponent>
));

const mockedGetProductRelationsAction = jest.fn();
jest.mock('@shopgate/pwa-common-commerce/product/actions/fetchProductRelations', () => (...args) => {
  mockedGetProductRelationsAction(...args);
  return {
    type: 'action',
  };
});

const mockedRelatedProducts = [];
const mockedProductRelations = [];
jest.mock('@shopgate/pwa-common-commerce/product/selectors/relations', () => ({
  getRelatedProducts: () => () => mockedRelatedProducts,
  getProductRelations: () => () => mockedProductRelations,
}));

jest.mock('@shopgate-ps/pwa-extension-kit/env/helpers', () => ({
  isIOSTheme: () => false,
}));

describe('Slider', () => {
  /**
   * Makes a component.
   * @returns {Object}
   */
  const makeComponent = () => mount((
    <Provider store={mockedStore({})}>
      <Slider
        productId="mockedId"
        type="mockedType"
        showPrice
        showName
      />
    </Provider>
  ));
  it('should call action on mount and render nothing', () => {
    const component = makeComponent();

    expect(mockedGetProductRelationsAction).toHaveBeenCalledWith({
      productId: 'mockedId',
      type: 'mockedType',
    });
    expect(component.html()).toBe('');
  });

  it('should render a slider', () => {
    mockedRelatedProducts.push({ id: 'mockedRelatedId' });
    mockedProductRelations.push('mockedRelatedId');
    const component = makeComponent();

    expect(component.find('MockedDummyComponent').props()).toMatchObject({
      products: { mockedRelatedId: { id: 'mockedRelatedId' } },
      showPrice: true,
      showName: true,
    });
    expect(component.html()).toMatchSnapshot();
  });
});
