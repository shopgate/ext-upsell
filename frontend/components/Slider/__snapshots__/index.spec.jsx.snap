// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render a slider 1`] = `"<div class=\\"css-1v3lwar\\"><div>Mocked Slider</div></div>"`;
